api/src/syscall/ipc/msg.rs

```diff
@@ -0,0 +1,593 @@
+    let is_privileged = current_uid == 0; // root user check
+
+    // Validate command code
+    if cmd != IPC_STAT
+        && cmd != IPC_SET
+        && cmd != IPC_RMID
+        && cmd != IPC_INFO
+        && cmd != MSG_INFO
+        && cmd != MSG_STAT
+    {
+        // Simplified: do not support some Linux extensions
+        return Err(AxError::from(LinuxError::EINVAL)); // EINVAL
+    }
+
+    // IPC_INFO (put before looking up the queue!)
+    if cmd == IPC_INFO {
+        // IPC_INFO uses msqid=0, no actual queue needed
+        // Return system-level information
+        #[repr(C)]
+        struct MsgInfo {
+            msgpool: i32,
+            msgmap: i32,
+            msgmax: i32,
+            msgmnb: i32,
+            msgmni: i32,
+            msgssz: i32,
+            msgtql: i32,
+            msgseg: u16,
+        }
+
+        let info = MsgInfo {
+            msgpool: 0,
+            msgmap: 0,
+            msgmax: MSGMAX as i32,
+            msgmnb: MSGMNB as i32,
+            msgmni: MSGMNI as i32,
+            msgssz: 0,
+            msgtql: 0,
+            msgseg: 0,
+        };
+
+        // Copy to user space
+        let ptr = buf as *mut MsgInfo;
+        ptr.vm_write(info)?;
+        return Ok(0);
+    }
+
+    // MSG_INFO (put before looking up the queue!)
+    if cmd == MSG_INFO {
+        let msg_manager = MSG_MANAGER.lock();
+        // Manually create IpcPerm
+        let msg_perm = starry_core::shm::IpcPerm {
+            key: 0,
+            uid: current_uid,
+            gid: current_gid,
+            cuid: current_uid,
+            cgid: current_gid,
+            mode: 0o600,
+            pad: 0,
+            seq: 0,
+            unused0: 0,
+            unused1: 0,
+        };
+
+        // Create a temporary msqid_ds to return information
+        let info_ds = msqid_ds {
+            msg_perm,
+            msg_stime: 0,
+            msg_rtime: 0,
+            msg_ctime: 0,
+            msg_cbytes: msg_manager.total_bytes() as u64,
+            // Use msg_qnum to return the number of allocated queues
+            msg_qnum: msg_manager.queue_count() as u64,
+            // Use msg_qbytes to return system limits or usage
+            msg_qbytes: MSGMNB as u64,
+            msg_lspid: Pid::from(0u32) as _,
+            msg_lrpid: Pid::from(0u32) as _,
+        };
+
+        // Copy to user space
+        let ptr = buf as *mut msqid_ds;
+        ptr.vm_write(info_ds)?;
+
+        // Return the current number of allocated queues
+        return Ok(msg_manager.queue_count() as isize);
+    }
+    // MSG_STAT handling
+    if cmd == MSG_STAT {
+        let msg_manager = MSG_MANAGER.lock();
+
+        let result = msg_manager
+            .iter_active_queues()
+            .nth(msqid as usize)
+            .and_then(|(actual_msqid, queue)| {
+                let guard = queue.lock();
+                let ptr = buf as *mut msqid_ds;
+                ptr.vm_write(guard.msqid_ds).ok()?;
```

Comment:
Error information might be lost here. Consider transform the return value of `.nth(...)` from an `Option` into a `Result` first and then call `Result::and_then` to preserve the error info.